### PR TITLE
Add GreenBench implementation

### PR DESCRIPTION
**Pull request by jiradeto**
This PR ports implementation of https://github.com/Rigorous-Software-Engineering/greenbench into fuzzbench. Essentially, we introduce new experiment option to allow users to opt in micro benchmark fuzzing that described in the GreenBench paper.

### CODE REVIEW

File: experiment/runner.py [hunk 1]
```diff
@@ -257,10 +268,13 @@ def set_up_corpus_directories(self):
             FUZZ_TARGET_DIR, fuzz_target_name)
         input_corpus = environment.get('SEED_CORPUS_DIR')
         os.makedirs(input_corpus, exist_ok=True)
-        if not environment.get('CUSTOM_SEED_CORPUS_DIR'):
-            _unpack_clusterfuzz_seed_corpus(target_binary, input_corpus)
+        if environment.get('MICRO_EXPERIMENT'):
+            _unpack_random_corpus(input_corpus)
         else:
-            _copy_custom_seed_corpus(input_corpus)
+            if not environment.get('CUSTOM_SEED_CORPUS_DIR'):
```
Comment (jonathanmetzman) [inline experiment/runner.py:274]: can you make this elif..else instead of nesting.

File: experiment/scheduler.py [hunk 2]
```diff
@@ -701,6 +701,7 @@ def __init__(self, trial):
         self.time_ended = trial.time_ended
         self.preemptible = trial.preemptible
         self.cpuset = None
+        self.trial_group_num = trial.trial_group_num
```
Comment (jonathanmetzman) [inline experiment/scheduler.py:704]: I think this has been plumbed through properly but I'm not 100% sure (no reason to think it's wrong). Sorry this setup isn't great.

File: experiment/runner.py [hunk 3]
```diff
@@ -100,6 +100,17 @@ def get_clusterfuzz_seed_corpus_path(fuzz_target_path):
...
     return seed_corpus_path if os.path.exists(seed_corpus_path) else None
 
+def _unpack_random_corpus(corpus_directory):
+    shutil.rmtree(corpus_directory)
+
+    benchmark = environment.get('BENCHMARK')
+    trial_group_num = 0
+    trial_group_num = environment.get('TRIAL_GROUP_NUM') 
+    random_corpora_dir = experiment_utils.get_random_corpora_filestore_path()
+    random_corpora_sub_dir = 'trial-group-%s' % int(trial_group_num)
```
Comment (jonathanmetzman) [inline experiment/runner.py:110]: can you use an f-string?

File: experiment/runner.py [hunk 4]
```diff
@@ -100,6 +100,17 @@ def get_clusterfuzz_seed_corpus_path(fuzz_target_path):
                         SEED_CORPUS_ARCHIVE_SUFFIX)
     return seed_corpus_path if os.path.exists(seed_corpus_path) else None
 
+def _unpack_random_corpus(corpus_directory):
+    shutil.rmtree(corpus_directory)
+
+    benchmark = environment.get('BENCHMARK')
+    trial_group_num = 0
+    trial_group_num = environment.get('TRIAL_GROUP_NUM')
```
Comment (jonathanmetzman) [inline experiment/runner.py:108]: environment.get('TRIAL_GROUP_NUM', 0)

File: common/random_corpus_fuzzing_utils.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
+"""Utility function for micro-experiment run."""
```
Comment (jonathanmetzman) [inline common/random_corpus_fuzzing_utils.py:1]: functions.

File: common/random_corpus_fuzzing_utils.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    with multiprocessing.Pool(*pool_args) as pool:
+        pool.starmap(prepare_benchmark_random_corpus, [
+            (benchmark, num_trials) for benchmark in benchmarks
+        ])
+        logs.info('Done Preparing corpus for micro experiment')
+
+def prepare_benchmark_random_corpus(benchmark: str,
+                                    num_trials: int):
+    
+    # temporary location to park corpus files before get picked randomly
```
Comment (jonathanmetzman) [inline common/random_corpus_fuzzing_utils.py:30]: Please start comments with a capital letter and end with a period.

File: common/random_corpus_fuzzing_utils.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+def initialize_random_corpus_fuzzing(benchmarks: List[str],
+                                     num_trials: int):
+    """Prepare corpus for micro experiment."""
+    pool_args = ()
+    with multiprocessing.Pool(*pool_args) as pool:
+        pool.starmap(prepare_benchmark_random_corpus, [
+            (benchmark, num_trials) for benchmark in benchmarks
+        ])
+        logs.info('Done Preparing corpus for micro experiment')
```
Comment (jonathanmetzman) [inline common/random_corpus_fuzzing_utils.py:25]: lowecase p in Preparing.

**Review by jonathanmetzman — COMMENTED** [review]
This looks pretty good, and I think it will be easy to merge this next week. I still need to take a deeper look at random_corpus_fuzzing_utils.py though.
Please fix the presubmit failure btw.

**Comment by jonathanmetzman** [comment]
Also could you write short documentation explaining how to use the feature, I want to play with it.

File: experiment/scheduler.py (hunk 2 above)
Comment (jiradeto) [reply experiment/scheduler.py:704]: Hi @jonathanmetzman the `trial_group_num` field is used as an identifier of the group of trials and this idea of trial grouping is used to make sure that we provide the same set of seeds to certain trails (within the same group).

For example, if in the config.yaml we set trials to 2 for the experiment so we have following trials in total (for fuzzer afl, libfuzzer and benchmark libjpeg):
- trial_id: 1, benchmark: libjpeg, fuzzer: afl
- trial_id: 2, benchmark: libjpeg, fuzzer: afl
- trial_id: 3, benchmark: libjpeg, fuzzer: libfuzzer
- trial_id: 4, benchmark: libjpeg, fuzzer: libfuzzer

so fuzzbench will generate following group (synthetically) and assign the same set initial seeds to trials within the same group:
- trial_id: 1, benchmark: libjpeg, fuzzer: afl   --> Group A
- trial_id: 2, benchmark: libjpeg, fuzzer: afl   --> Group B
- trial_id: 3, benchmark: libjpeg, fuzzer: libfuzzer --> Group A
- trial_id: 4, benchmark: libjpeg, fuzzer: libfuzzer --> Group B

Please let me know if it makes sense? also is there a way to only add/inject this `trial_group_num` field only when are running in `MICRO_EXPERIMENT` mode?

**Comment by jiradeto** [comment]
> Also could you write short documentation explaining how to use the feature, I want to play with it.

hi @jonathanmetzman, as this feature depends on oss-fuzz-corpus, so we need to specify `--oss-fuzz-corpus` flag when starting experiment and also set new parameter:
```
micro_experiment: true
```
in experiment-config.yml file.

**Comment by jonathanmetzman** [comment]
Can you fix the test failure please?

**Comment by jiradeto** [comment]
Fixed all failed tests @jonathanmetzman can you please help to re-review?

**Review by jonathanmetzman — APPROVED** [review]
LGTM

**Pull request merged**